include/pybind11/detail/internals.h

```diff
@@ -18,6 +18,27 @@ inline PyTypeObject *make_static_property_type();
 inline PyTypeObject *make_default_metaclass();
 inline PyObject *make_object_base_type(PyTypeObject *metaclass);
 
+// The old Python Thread Local Storage (TLS) API is deprecated in Python 3.7 in favor of the new
+// Thread Specific Storage (TSS) API.
+#if PY_VERSION_HEX >= 0x03070000
+    #define PYBIND11_TLS_KEY_TYPE Py_tss_t *
+    #define PYBIND11_TLS_KEY_NULL_VALUE nullptr
+    #define PYBIND11_TLS_GET_VALUE(key) PyThread_tss_get((key))
+    #define PYBIND11_TLS_REPLACE_VALUE(key, value) PyThread_tss_set((key), (tstate))
+    #define PYBIND11_TLS_DELETE_VALUE(key) PyThread_tss_set((key), nullptr)
+#else
+    // Usually an int but a long on Cygwin64 with Python 3.x
+    #define PYBIND11_TLS_KEY_TYPE decltype(PyThread_create_key())
+    #define PYBIND11_TLS_KEY_NULL_VALUE 0
+    #define PYBIND11_TLS_GET_VALUE(key) PyThread_get_key_value((key))
+    #if PY_MAJOR_VERSION < 3
+        #define PYBIND11_TLS_REPLACE_VALUE(key, value) do { PyThread_delete_key_value((key)); PyThread_set_key_value((key), (value)); } while (false)
+    #else
+        #define PYBIND11_TLS_REPLACE_VALUE(key, value) PyThread_set_key_value((key), (value))
+    #endif
+    #define PYBIND11_TLS_DELETE_VALUE(key) PyThread_set_key_value((key), nullptr)
```

Comment:
We should use `PyThread_delete_key_value` for `PYBIND11_TLS_DELETE_VALUE`.

`PyThread_set_key_value` is strange that it does not change key value to nullptr when it's already set to non-null value, tested on Python 2.7.12.

I notice this because my Python callback function (overriding virtual function by `PYBIND11_OVERLOAD_PURE`) always segfaults upon the second call.

Reply:
@cyfdecyf And did it do that before this PR? Because I thought the code from before and after would be the same, on < 3.7 Python versions?

Reply:
Hmmm, you're right, something weird is happening, there, using this macro :|

Reply:
@YannickJadoul, @cyfdecyf : could you check if PR #1517 fixes the issue?

Reply:
Sorry for the late reply. I'll try to construct a test case for this issue.

Reply:
@wjakob I can confirm your patch fixes a regression we're seeing in PyTorch https://github.com/pytorch/pytorch/pull/11473#issuecomment-420155829. We have a release this Friday, do you have an ETA on when that fix can be merged? Thanks!

Reply:
I will merge it today and push out a new patch release today.